functional_tests.go

```diff
@@ -255,13 +266,19 @@ func isQuickMode() bool {
 	return os.Getenv("MODE") == "quick"
 }
 
+func getFuncName() string {
+	pc, _, _, _ := runtime.Caller(1)
+	return strings.TrimPrefix(runtime.FuncForPC(pc).Name(), "main.")
+}
+
 // Tests bucket re-create errors.
 func testMakeBucketError() {
 	region := "eu-central-1"
 
 	// initialize logging params
 	startTime := time.Now()
-	function := "MakeBucket(bucketName, region)"
+	testName := getFuncName()
+	function := []string{"MakeBucket(bucketName, region)"}
```

Comment:
Isn't each test suppose to test one SDK API primarily, even though it may use many API methods? Why do we need a `[]string` for `function` field in the log json? This is different from our decided format here - https://github.com/minio/mint#mint-log-format

Reply:
That's true, but there are still some test cases already part of the existing code, like `testFunctionalV2`, and `testFunctional`, which come as a bunch of api tests. I guess these tests have been added as part  of the new `QuickTest` mode which is why the string array is introduced recently.

Reply:
Does it mean you need to fix the test than breaking JSON?

Reply:
No, we don't need to fix anything, imho. That is  perfectly fine.
So, what I did was to accommodate these kind of test cases and handle them in a kind of a better way.
Now, when these tests run, each and every api name will be printed in the `function` field as part of the json file if there is an error. If all of the api tests pass, then all of the api names will be printed out in the `function` field as part of the json file. So, no problem.

Reply:
Can you provide sample JSONs on success and failure cases?

Reply:
Success example:
```
go build functional_tests.go;go run functional_tests.go | jq
{
  "args": {
    "bucketName": "minio-go-test-hx5t3l96fruilwyi"
  },
  "duration": 2119,
  "function": [
    "MakeBucket(bucketName, region)",
    "BucketExists(bucketName)",
    "GetBucketPolicy(bucketName, objectPrefix)",
    "SetBucketPolicy(bucketName, objectPrefix, bucketPolicy)",
    "ListBuckets()",
    "PutObject(bucketName, objectName, reader, contentType)",
    "ListObjects(bucketName, objectName, isRecursive, doneCh)",
    "ListObjectsV2(bucketName, objectName, isRecursive, doneCh)",
    "ListIncompleteUploads(bucketName, objectName, isRecursive, doneCh)",
    "GetObject(bucketName, objectName)",
    "FGetObject(bucketName, objectName, fileName)",
    "PresignedHeadObject(bucketName, objectName, expires, reqParams)",
    "PresignedGetObject(bucketName, objectName, expires, reqParams)",
    "PresignedPutObject(bucketName, objectName, expires)",
    "RemoveObject(bucketName, objectName)",
    "RemoveBucket(bucketName)"
  ],
  "name": "minio-go: testFunctional",
  "status": "PASS"
}
{
  "args": {
    "bucketName": "minio-go-test-04ihrp4hwhtm2dst",
    "objectName": "minio-go-test-04ihrp4hwhtm2dstunique-presigned"
  },
  "duration": 634,
  "function": [
    "MakeBucket(bucketName, region)",
    "BucketExists(bucketName)",
    "SetBucketPolicy(bucketName, objectPrefix, bucketPolicy)",
    "ListBuckets()",
    "ListObjects(bucketName, objectName, isRecursive, doneCh)",
    "ListIncompleteUploads(bucketName, objectName, isRecursive, doneCh)",
    "GetObject(bucketName, objectName)",
    "FGetObject(bucketName, objectName, fileName)",
    "PresignedHeadObject(bucketName, objectName, expires, reqParams)",
    "PresignedGetObject(bucketName, objectName, expires, reqParams)",
    "PresignedPutObject(bucketName, objectName, expires)"
  ],
  "name": "minio-go: testFunctionalV2",
  "status": "PASS"
}
```

Failure example (induced an error just before SetBucketPolicy api execution):
```
$ go build functional_tests.go;go run functional_tests.go | jq
{
  "args": {
    "bucketName": "minio-go-test-eomoglxiubs5s9h9",
    "bucketPolicy": "readonly",
    "objectPrefix": ""
  },
  "duration": 169,
  "error": "BOOM!",
  "function": [
    "SetBucketPolicy(bucketName, objectPrefix, bucketPolicy)"
  ],
  "message": "SetBucketPolicy failed",
  "name": "minio-go: testFunctional",
  "status": "FAIL"
}
```

Reply:
I see you change `function` field as array which leads to dynamic JSON i.e. other test have the field as string (as per currently agreed JSON format).  You would need to fix tests, not JSON.

Reply:
@balamurugana 
I guess, the issue here is the way `quick` mode feature is designed and implemented. It is a break in the rule of having a single api tested in a single test case. Having a group of test cases running inside a wrapper and reporting the wrapper without the covered api names is a shortage of information that is supposed to be conveyed to the end user. That's what is behind the code change of having `function` field transformed from `string` to `array` type. So, it was an attempt to accommodate the existing `quick` mode feature design and implementation to our existing standard with minimum effort.

Now, if we don't want to change the `function` field to `array` type, then we need to decide if we are ok with not reporting apis tested under `testFunctionalV2`, and `testFunctional` test cases.

Please let me  know what you think.

We also need to make sure our whole mint team understand that grouping of api tests in one single test case is against our mint testing design philosophy.

Reply:
@balamurugana 
(It is not the finest looking code, but) I kept the `function` field as `string` and populated it with the list of all apis tested.  
Now here how it loooks like:

Success:
```
$ go build functional_tests.go;go run functional_tests.go | jq
{
  "args": {
    "bucketName": "minio-go-test-lj99jvecgld2v1rk"
  },
  "duration": 1486,
  "function": "MakeBucket(bucketName, region), BucketExists(bucketName), GetBucketPolicy(bucketName, objectPrefix), SetBucketPolicy(bucketName, objectPrefix, bucketPolicy), GetBucketPolicy(bucketName, objectPrefix), SetBucketPolicy(bucketName, objectPrefix, bucketPolicy), GetBucketPolicy(bucketName, objectPrefix), SetBucketPolicy(bucketName, objectPrefix, bucketPolicy), GetBucketPolicy(bucketName, objectPrefix), ListBuckets(), PutObject(bucketName, objectName, reader, contentType), ListObjects(bucketName, objectName, isRecursive, doneCh), ListObjectsV2(bucketName, objectName, isRecursive, doneCh), ListIncompleteUploads(bucketName, objectName, isRecursive, doneCh), GetObject(bucketName, objectName), FGetObject(bucketName, objectName, fileName), PresignedHeadObject(bucketName, objectName, expires, reqParams), PresignedHeadObject(bucketName, objectName, expires, reqParams), PresignedGetObject(bucketName, objectName, expires, reqParams), PresignedGetObject(bucketName, objectName, expires, reqParams), PresignedPutObject(bucketName, objectName, expires), PresignedPutObject(bucketName, objectName, expires), RemoveObject(bucketName, objectName), RemoveBucket(bucketName)",
  "name": "minio-go: testFunctional",
  "status": "PASS"
}
{
  "args": {
    "bucketName": "minio-go-test-oimq6as19p2ut49e",
    "objectName": "minio-go-test-oimq6as19p2ut49eunique-presigned"
  },
  "duration": 437,
  "function": "MakeBucket(bucketName, location), BucketExists(bucketName), SetBucketPolicy(bucketName, objectPrefix, bucketPolicy), ListBuckets(), ListObjects(bucketName, objectName, isRecursive, doneCh), ListIncompleteUploads(bucketName, objectName, isRecursive, doneCh), GetObject(bucketName, objectName), FGetObject(bucketName, objectName, fileName), PresignedHeadObject(bucketName, objectName, expires, reqParams), PresignedGetObject(bucketName, objectName, expires, reqParams), PresignedPutObject(bucketName, objectName, expires), GetObject(bucketName, objectName)",
  "name": "minio-go: testFunctionalV2",
  "status": "PASS"
}
```

Failure:
```
go build functional_tests.go;go run functional_tests.go | jq
{
  "args": {
    "bucketName": "minio-go-test-w3i5rbgfrvh2j5xe"
  },
  "duration": 212,
  "error": "BOOM!",
  "function": "BucketExists(bucketName)",
  "message": "BucketExists failed",
  "name": "minio-go: testFunctional",
  "status": "FAIL"
}
```

Reply:
I don't know whether I understood `quick` mode design/implementation here.   All tests in minio-go are run in all modes i.e. there is no difference for `quick` and `enhanced` modes.

I haven't completely understood any arguments about mint JSON till now, hence I can't comment on that. I am only interested in whether I will find what/where the failure happened in tests so that I find the problem and fix it accordingly.  If failure log provides such information, I am OK.

If you say below JSON helps tracing down the problem whether its in test and/or server, its good to go.
```json
{
  "args": {
    "bucketName": "minio-go-test-w3i5rbgfrvh2j5xe"
  },
  "duration": 212,
  "error": "BOOM!",
  "function": "BucketExists(bucketName)",
  "message": "BucketExists failed",
  "name": "minio-go: testFunctional",
  "status": "FAIL"
}
```